_posts/2021-11-02-choosing-dask-chunk-sizes.md

```diff
@@ -0,0 +1,157 @@
+
+If you are reading data from disk, the storage structure will inform what shape your Dask array chunks should be. For best performance, choose chunks that are well aligned with the way data is stored.
+
+From the Dask best practices on how to [orient your chunks](https://docs.dask.org/en/stable/array-best-practices.html#orient-your-chunks):
+
+> When reading data you should align your chunks with your storage format. Most array storage formats store data in chunks themselves. If your Dask array chunks aren’t multiples of these chunk shapes then you will have to read the same data repeatedly, which can be expensive. Note though that often storage formats choose chunk sizes that are much smaller than is ideal for Dask, closer to 1MB than 100MB. In these cases you should choose a Dask chunk size that aligns with the storage chunk size and that every Dask chunk dimension is a multiple of the storage chunk dimension.
+
+Some examples of data storage structures on disk include:
+- A HDF5 or [Zarr array](https://zarr.readthedocs.io/en/stable/api/core.html). The size and shape of chunks/blocks stored on disk should align well with the Dask array chunks you select.
+- A folder full of tiff files. You might decide that each tiff file should become a single chunk in the Dask array (or that multiple tiff files should be grouped into a single chunk).
+
+### Rechunking arrays
+
+If you want to change the chunking of a Dask array, you can do that with the [rechunk](https://docs.dask.org/en/latest/generated/dask.array.rechunk.html) method.
+
+```python
+rechunked_array = original_array.rechunk(new_shape)
+```
+
+**Warning:** Rechunking Dask arrays comes at a cost.
+- The Dask graph must be rearranged to accomodate the new chunk structure. This happens immediately, and will block any other interaction with python until Dask has rearranged the task graph.
+- This also inserts new tasks into the Dask graph. At compute time, there are now more tasks to execute.
+
+For these reasons, it is best to avoid rechunking. However, sometimes the data is stored on disk is not well aligned and rechunking may be necessary.
+For an example of this, here is Draga Doncila Pop [talking about chunk alignment](https://youtu.be/10Ws59NGDaE?t=833) with satellite image data.
+
+## Too small is a problem
+
+If array chunks are too small, it's inefficient. Why is this?
+
+Using Dask introduces some amount of overhead for each task in your computation.
+This overhead is the reason the Dask best practices advise you to [avoid too-large graphs](https://docs.dask.org/en/latest/best-practices.html#avoid-very-large-graphs).
+This is because if the amount of actual work done by each task is very tiny, then the percentage of overhead time vs useful work time is not good.
+
+Typically, the Dask scheduler takes 1 millisecond to coordinate a single task. That means we want the computation time for each task to be comparitively larger, eg: seconds instead of milliseconds.
```

Comment:
This is already said above. But I would remove the other one and keep this one.

Reply:
I agree that this is important to keep here (and now with the rearranging of sections, this part comes first in the article). 

But I don't want to remove it from the rules of thumb list for two reasons:
1 It is a very clear and specific guideline so I think it belongs, and
2. Most people will assume the "rough rules of thumb" list is a good summary of the takeaways from this article, so if we omit it from the list people will likely forget or overlook this point.
I think a small amount of repetition is ok in this case.